pkg/node/manager/manager.go

```diff
@@ -339,7 +340,11 @@ func (m *Manager) NodeUpdated(n nodeTypes.Node) {
 
 	for _, address := range n.IPAddresses {
 		var tunnelIP net.IP
-		if address.Type == addressing.NodeCiliumInternalIP || m.conf.NodeEncryptionEnabled() {
+		// If the host firewall is enabled, all traffic to remote nodes must go
+		// through the tunnel to preserve the source identity as part of the
+		// encapsulation.
+		if address.Type == addressing.NodeCiliumInternalIP || m.conf.NodeEncryptionEnabled() ||
+			option.Config.EnableHostFirewall {
```

Comment:
Yep, we need to send through tunnel because otherwise there's some SNAT happening and we lose the source security ID. I'll update the commit message.

Is it possible/sensible for users to run in tunneling mode even though direct routing to remote nodes works? I kind of expected that traffic from pods to remote nodes would be SNATed all the time; otherwise, why use a tunnel?

Reply:
:+1: good to declare assumptions.

I guess I'd come from a perspective of "if you run a tunnel you know it 'works' anywhere regardless of the underlying fabric" and that "SNAT should be done as little as possible" but yeah probably if the user knows they're running all nodes in the same L2 or if they know that their underlying L3 fabric provides routing for the addresses (or run something like kube-router / metallb / bird for advertising the pod prefix) then they should just run direct routing. And I think your expectation is probably fair, if we don't have these prior guarantees then SNAT is the best bet we have to ensure that traffic arrives at a destination.

Anyway I think just documenting this in the commits/code and filing an issue for any caveats/remaining bugs is fine to address this in the context of this PR.